lamagraph-compiler/src/Lamagraph/Compiler/Nets/Reduction.hs

```diff
@@ -60,23 +63,31 @@ reduceStep rule conf@Configuration{stack, cycles, threadState} = case threadStat
   -- T.1
   Delist -> case stack of
     [] -> throwIO $ DelistFromEmptyStackException conf
-    hd : tl -> pure $ conf{stack = tl, threadState = Process hd}
-  -- T.2
-  Enlist (hd : tl) -> pure $ conf{stack = hd : stack, threadState = Enlist tl}
-  -- T.3
-  Enlist [] -> pure $ conf{threadState = Delist}
+    [Nothing] -> throwIO $ DelistFromEmptyStackException conf
+    hd : tl -> case hd of
+      Just pair -> pure $ conf{stack = tl, threadState = Process pair}
+      Nothing -> do
+        env <- ask
+        atomically $ do
+          currentReductionCounter <- readTVar $ getReductionCounter env
+          prevReductionCounter <- readTVar $ getPreviousReductionCounter env
+          modifyTVar' (getReductionWidthHistory env) ((currentReductionCounter - prevReductionCounter) :)
+          writeTVar (getPreviousReductionCounter env) currentReductionCounter
+        pure $ conf{stack = tl <> [Nothing]}
+  -- T.2 + T.3
+  Enlist pairs -> pure $ conf{stack = stack <> (Just <$> pairs), threadState = Delist}
   -- T.4
-  Cycle newCycle -> pure $ conf{cycles = uncurry Map.insert newCycle cycles}
+  Cycle newCycle -> pure $ conf{cycles = uncurry Map.insert newCycle cycles, threadState = Delist}
   -- Main processing
   Process pair -> process rule conf pair
 
--- | Looping version of 'reduceStep'. Works until 'stack' is empty and 'threadState' is 'Delist'.
+-- | Looping version of 'reduceStep'. Works until 'stack' is empty (actually contains only 'Nothing') and 'threadState' is 'Delist'.
```

Comment:
Wow. Why empty stack is not just ```[]```?

Reply:
Added documentation why stack now works this way